Sources/C++Workshop/Cards_helper.h

```diff
@@ -14,20 +14,33 @@ template <typename Deck> class Card;
 template <typename Impl>
 class Deck {
 protected:
-    using Cards = std::vector<std::reference_wrapper<Card<Impl>>>;
+    using Card = Card<Impl>;
+    using Cards = std::vector<std::unique_ptr<Card>>;
     Cards _cards;
     
+    static auto copy_cards(const Cards& cards) {
+        auto out = Cards{};
+        std::transform(cards.begin(), cards.end(), std::back_inserter(out), [](auto& card) {
+            return std::make_unique<Card>(*card);
+        });
+        return out;
+    }
+    
 public:
     using size_type = typename decltype(_cards)::size_type;
     using iterator = typename decltype(_cards)::iterator;
     
-    Deck(const Cards& cards): _cards{cards} { }
-    Deck(const Deck&) = default;
+    Deck(Cards&& cards): _cards{std::move(cards)} { }
+    Deck(const Cards& cards): _cards{copy_cards(cards)} { }
+    Deck(const Deck& other): Deck{copy_cards(other._cards)} { }
+    auto operator=(const Deck& other) {
+        this->_cards = copy_cards(other._cards);
+    }
     
     auto empty() const { return this->_cards.empty(); }
     auto size() const { return this->_cards.size(); };
     
-    auto at(size_type index) const { return this->_cards.at(index); }
+    auto at(size_type index) const -> const typename Cards::value_type& { return this->_cards.at(index); }
```

Comment:
I'm curious if this is required or just nice to have...?

Reply:
I'm basically just using `value_type` as a shortcut to `unique_ptr<Card>` (and if I were to change `Cards` in the future, these return types would also change).